pmd-kotlin/src/main/java/net/sourceforge/pmd/cpd/KotlinTokenizer.java

```diff
@@ -0,0 +1,104 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.cpd;
+
+import org.antlr.v4.runtime.ANTLRInputStream;
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.Lexer;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.Token;
+
+import net.sourceforge.pmd.lang.ast.TokenMgrError;
+import net.sourceforge.pmd.lang.kotlin.antlr4.Kotlin;
+
+/**
+ * The Kotlin Tokenizer
+ */
+public class KotlinTokenizer implements Tokenizer {
+
+    private boolean discardingPackageAndImport = false;
+
+    @Override
+    public void tokenize(final SourceCode sourceCode, final Tokens tokenEntries) {
+        final StringBuilder buffer = sourceCode.getCodeBuffer();
+
+        try {
+            final ANTLRInputStream ais = new ANTLRInputStream(buffer.toString());
+            final Kotlin lexer = new Kotlin(ais);
+
+            lexer.removeErrorListeners();
+            lexer.addErrorListener(new ErrorHandler());
+            Token token = lexer.nextToken();
+
+            while (token.getType() != Token.EOF) {
+                analyzeTokenStart(token);
+                if (token.getChannel() != Lexer.HIDDEN && token.getType() != Kotlin.NL && !isDiscarding()) {
+                    final TokenEntry tokenEntry = new TokenEntry(token.getText(), sourceCode.getFileName(), token.getLine());
+                    tokenEntries.add(tokenEntry);
+                }
+                analyzeTokenEnd(token);
+                token = lexer.nextToken();
+            }
+        } catch (final ANTLRSyntaxError err) {
+            // Wrap exceptions of the Kotlin tokenizer in a TokenMgrError, so
+            // they are correctly handled
+            // when CPD is executed with the '--skipLexicalErrors' command line
+            // option
+            throw new TokenMgrError("Lexical error in file " + sourceCode.getFileName() + " at line " + err.getLine()
+                    + ", column " + err.getColumn() + ".  Encountered: " + err.getMessage(),
+                    TokenMgrError.LEXICAL_ERROR);
+        } finally {
+            tokenEntries.add(TokenEntry.getEOF());
+        }
+    }
+
+    private boolean isDiscarding() {
```

Comment:
all filtering / discarding should be done through a `net.sourceforge.pmd.cpd.token.TokenFilter`, as is done in many other languages.